Turn QueueCPU* into template QueueGeneric*<TDev>

jkelling opened the pull request:
This patch turns QueueCpuBlocking and QueueCpuNonBlocking into template with the Device type as parameter.

### Motivation
This will allow reusing the blocking and non-blocking CPU queue implementations for new backend, which do not require their own queue implementation but which do need a specific device class to abstract target devices. At the moment candidates are the [OpenMP 5 target-offload backend](https://github.com/jkelling/alpaka) and OpenACC.

### Main Changes
* add templates QueueGenericBlocking, QueueGenericNonBlocking, EventGeneric, IGenericQueue (using the code of their CPU predecessors, except with TDev instead of DevCpu)
* add aliases QueueCpuBlocking, QueueCpuNonBlocking, EventCpu, ICpuQueue, maintaning the old file structure, to limit the impact on other part of the code using the old classes
* replace forward declarations of EventCpu in some tests with includes as they would collide with the alias
* add new traits RegisterQueue and GetAllQueues for DevCpu

jkelling committed:
Add QueueGeneric*Blocking

Add QueueCpu*Blocking = using QueueGeneric*Blocking<DevCpu, ...>

jkelling committed:
QueueGeneric*Blocking: Add IGenericQueue

Rm ICpuQueue

jkelling committed:
dev::traits: Add RegisterQueue, GetAllQueues

QueueGeneric*: use trait to register
EventGeneric: use trait to get all queues

jkelling committed:
Add ICpuQueue alias; Fix EventCpu includes

Fix QueueCpuOmp2Collective.hpp: include <memory>

jkelling committed:
QueueCpuÜ. EvenCpu: Rm concrete impl

BenjaminW3 commented on include/alpaka/queue/IGenericQueue.hpp:
```diff
@@ -0,0 +1,57 @@
+#include <alpaka/core/BoostPredef.hpp>
+
+// #include <alpaka/event/EventGeneric.hpp>
+// #include <alpaka/dev/DevCpu.hpp>
```
Remove when not needed?

BenjaminW3 commented on include/alpaka/queue/IGenericQueue.hpp:
```diff
@@ -0,0 +1,57 @@
+        //#############################################################################
+        //! The CPU queue interface
+        template<
+            typename TDev>
```
Could we make `IGenericQueue` depend on `TEvent` instead of `TDev`? The queue does not need to know its device at all but only the Event. `event::EventGeneric<TDev>` can be provided at the place the template is instantiated.

jkelling replied:
I did it this way to ensure the template argument will always be derived from EventGeneric and exclude arbitrary types. If we do not want to be this strict, we could use `template<typename TEvent>` here. What shall it be?

BenjaminW3 replied:
I do not see a reason to be so strict. Maybe there will be another queue that uses a totally different event.

jkelling replied:
But `IGenericThreadsQueue` is a polymorphic interface for `QueueThreadsGeneric*` which uses `EventGeneric<TDev>`.

Allowing `IGenericThreadsQueue` to be used with other queues than instances of `QueueThreadsGeneric*` would extent the original scope of `ICpuQueue`. If we extent the scope this far, `IGenericThreadsQueue` could also be used for something like the Cuda queue, and thus would be restricted to living on the host anymore.

jkelling replied:
At least this is what I understood the intent behind ICpuQueue to be: a polymorphic interface for queues using EventCpu.

BenjaminW3 replied:
I am also ok with this. My point was that there might be another type of event that can be used with the `IGenericThreadsQueue` that is still compatible with the current CPU queue and device. but is not necessarily a specialization of `EventGeneric`.

jkelling replied:
There can be no event which is not a specialization of `EventGenericThreads`, that can also be used with `QueueGenericThreats*` and by extension `QueueCpu*`, because the queue explictly uses `EventGenericThreads<TDev>` and also overrides `enqueue()` and `wait()` using this argument type. So if `IGenericThreadsQueue` were not restricted one of the following would need to change:
* formally allow `IGenericThreadsQueue` to be a base for different queue types
* add another template parameter `TEvent` to `QueueGeneric`, one could still enforce that `TEvent` and `TDev` are connected by saying
  ```c++
  template < typename TDev, template<typename TTDev> TEvent = EventGenericThreads >
  class QueueGenericThreats...
  ```
At the moment I would prefer to leave it strict. There is the option of relaxing it should the need arise, as `IGenericThreadsQueue` is an implementation detail from the perspective of user code.

BenjaminW3 commented on include/alpaka/queue/IGenericQueue.hpp:
```diff
@@ -0,0 +1,57 @@
+        //! The CPU queue interface
+        template<
+            typename TDev>
+        class IGenericQueue
```
I really do not like the `Generic` names for the queues and the event here. They are not that generic at all. The `Cpu` part of the name previously had two meanings:
* the queue holds `EventCpu`
* the queue is executed on the CPU

You made the first point generic but what it still differentiates from a generic queue or the CUDA queues is that it can only be run on the host CPU.
I would prefer if this would be called `IGenericCpuQueue` and still be located in the `cpu` subfolder. The same is true for the queue impls and the events.

jkelling replied:
OK, I thought the name was only related to the Dev / Event type.

Howver, the name `CPU` then is still too general in that it only says that it lives on the host. There are multiple ways to implement a device-wise generic queue. This on used C++11 threads, on different option would be OpenMP tasks. With this in mind, we could call them, e.g. `IGenericThreadsQueue`, `QueueThreadsBlocking`. The interface would still go into the cpu subfolder.

BenjaminW3 replied:
I like your proposed names

jkelling replied:
I meant to write `QueueGenericThreadsBlocking` instead of `QueueThreadsBlocking`, Is that ok, too?

BenjaminW3 replied:
Also fine

BenjaminW3 commented on include/alpaka/queue/QueueGenericNonBlocking.hpp:
```diff
@@ -0,0 +1,244 @@
+    }
+}
+
+#include <alpaka/event/EventGeneric.hpp>
```
This looks strange

jkelling replied:
This is just copied from the implementation of QueueCpuNonBlocking. I guess someone wanted to have the queue header also provide the event definition. The Event header cannot included at the top though, because it requires the definition of the queue for the trait specializations. I did not try removing it, assuming it may break other code.

BenjaminW3 replied:
I am fine with this

BenjaminW3 commented on include/alpaka/event/EventGeneric.hpp:
```diff
@@ -0,0 +1,508 @@
+        //! The CPU device event.
+        template<
+            typename TDev>
+        class EventGeneric final : public concepts::Implements<wait::ConceptCurrentThreadWaitFor, EventGeneric<TDev>>
```
As written in the other comment, I would really like to call this `EventCpuGeneric` and move it into a `cpu` subfolder.

psychocoderHPC replied:
~~The only problem I see with the name `EventCpuGeneric` is that the upcoming OpenMP5 and OpenAcc backends are not only for CPU.~~
[updated] The reason for the naming is [here](https://github.com/ComputationalRadiationPhysics/alpaka/pull/959/files#r395969995) better described

BenjaminW3 commented on include/alpaka/dev/Traits.hpp:
```diff
@@ -145,5 +159,40 @@ namespace alpaka
+            TDev const & dev)
+#ifdef BOOST_NO_CXX14_RETURN_TYPE_DEDUCTION
+        -> decltype(traits::GetAllQueues<TDev>::getAllQueues(dev))
+#endif
```
Please remove this ifdef as we are c++14 only now.

BenjaminW3 commented on include/alpaka/dev/DevCpu.hpp:
```diff
@@ -244,6 +244,43 @@ namespace alpaka
+                    dev::DevCpu const & dev)
+#ifdef BOOST_NO_CXX14_RETURN_TYPE_DEDUCTION
+                -> decltype(dev.m_spDevCpuImpl->GetAllQueues())
+#endif
```
please remove the ifdef

BenjaminW3 commented on include/alpaka/dev/Traits.hpp:
```diff
@@ -64,6 +64,20 @@ namespace alpaka
+            template<
+                typename T,
+                typename TSfinae = void>
+            struct GetAllQueues;
```
Do we need those public traits at all?
The `ICpuQueue` interface with the `registerQueue` and `getAllQueues` methods is an implementation detail of the CPU queues (and therefore lives in the `cpu` subfolder) and should not be publicly visible even after your changes. The CUDA queues do not implement this and do not need this.
It should be possible to live without those traits because the `QueueCpuGenericXxx` gets its specific device implementation as template parameter and the device simply has to implement a `RegisterQueue` method without the need for any explicit public interface.

jkelling replied:
So, it would be better to require TDev to implement some concept `QueueRegister`. I agree with skipping the traits.

jkelling replied:
Just one issue: DevCpu does not have `GetAllQueues()` and `RegisterQueue()`, they are only present in DevCpuImpl. In introduced the traits because it looked a it strange when the queue would just reach into the internal implementation of DevCpu. So how should this be solved?
1. Just access `dev.m_spDevOmp4Impl->RegisterQueue(queue.m_spQueueImpl)`.
2. Move the member functions to DevCpu, as they are part of its interface.

BenjaminW3 replied:
both sound reasonable so do what you think is easier. For the first solution you might get access issues and the members have to have the same name so the second might be easier.

jkelling replied:
I will do option 2 then, because it is cleaner, though  more work than 1.

BenjaminW3 reviewed (changes requested):
Thank your for providing the first steps to add the new OpenMP 4 support!

psychocoderHPC commented on include/alpaka/event/EventGeneric.hpp:
```diff
@@ -0,0 +1,508 @@
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto enqueue(
+#if !(BOOST_COMP_CLANG_CUDA && BOOST_ARCH_PTX)
+                    queue::generic::detail::QueueGenericNonBlockingImpl<TDev> & queueImpl,
```
please use `alpaka::ignore_unused(queueImpl)` in the function body.

```
#if BOOST_COMP_CLANG_CUDA && BOOST_ARCH_PTX
    alpaka::ignore_unused(queueImpl);
#endif
```

psychocoderHPC replied:
Could be that this code is derived from the original implementation but we should change it directly in this PR.

jkelling replied:
indeed

psychocoderHPC commented on include/alpaka/queue/QueueCpuBlocking.hpp:
```diff
@@ -11,7 +11,7 @@
 #include <alpaka/core/Unused.hpp>
 #include <alpaka/dev/DevCpu.hpp>
-#include <alpaka/queue/cpu/ICpuQueue.hpp>
+#include <alpaka/queue/QueueGenericBlocking.hpp>
```
all expect this include should be removed

jkelling replied:
I cannot follow, can you please clarify?

This file defines an alias for `QueueGenericBlocking<DevCpu>` and thus needs the definition of that template.

Do you mean this file should be removed and thus the alias? This would entail lots of changes elsewhere, i.e. `QueueCpuBlocking` -> `QueueGenericBlocking<DevCpu>`

BenjaminW3 replied:
I think @psychocoderHPC  meant that you can remove nearly all other includes in this file because you only create a typedef for a speialization of `QueueGenericBlocking`. So you only need its header and the header of the template parameter.
This is also true for the other typedef-only headers.

jkelling replied:
That does make a lot of sense, thanks. Done.

psychocoderHPC commented on include/alpaka/queue/QueueCpuNonBlocking.hpp:
```diff
@@ -10,7 +10,7 @@
 
 #include <alpaka/dev/DevCpu.hpp>
-#include <alpaka/queue/cpu/ICpuQueue.hpp>
+#include <alpaka/queue/QueueGenericNonBlocking.hpp>
```
all expect this include should be removed.

psychocoderHPC commented on include/alpaka/queue/QueueGenericNonBlocking.hpp:
```diff
@@ -0,0 +1,244 @@
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto enqueue(
+#if !(BOOST_COMP_CLANG_CUDA && BOOST_ARCH_PTX)
+                    queue::QueueGenericNonBlocking<TDev> & queue,
```
Please use `alpaka::ignore_unused()`

jkelling committed:
IGenericQueue: Rm inactive includes

jkelling committed:
{Queue,Event}Generic: style Fixes

jkelling committed:
DevCpu: Rm traits GetAllQueues, RegisterQueue

jkelling committed:
Rename QueueGeneric*Blocking -> QueueGenericThreads*Blocking

BenjaminW3 commented on include/alpaka/dev/DevCpu.hpp:
```diff
@@ -68,16 +68,30 @@ namespace alpaka
                 {
                 private:
 
+                public:
```
Could you please delete the `private:` section above?

jkelling committed:
DevCpuImpl: Rm empty private

jkelling committed:
QueueCpu* aliases: Rm unused includes

BenjaminW3 commented:
```
/home/travis/build/ComputationalRadiationPhysics/alpaka/include/alpaka/event/EventCpu.hpp(18): error #135: namespace "alpaka::dev" has no member "DevCpu"
```

jkelling commented:
> ```
> /home/travis/build/ComputationalRadiationPhysics/alpaka/include/alpaka/event/EventCpu.hpp(18): error #135: namespace "alpaka::dev" has no member "DevCpu"
> ```

So this is one of the two failing builds? I can never retrieve those build logs.
What is the cmake configuration in this case? What are the final compile flags? Is this running with `VERBOSE=1`?

jkelling commented:
I, see, I deleted one include too many and did not see the compiler error afterwards (or it build because of cached files somehow).

BenjaminW3 commented:
`ALPAKA_DEBUG=2`

jkelling committed:
EventCpu: Add include DevCpu

QueueCpu*: Move include of EventCpu to top

jkelling commented:
There is still one windows check failing and I cannot download the log. Any hints?

BenjaminW3 commented:
```
CMake Error at C:/Program Files/CMake/share/cmake-3.16/Modules/FindCUDA.cmake:1602 (configure_file):
configure_file Problem configuring file
```
I am pretty sure that this is not your fault but a temporary issue in CI.

psychocoderHPC commented:
@jkelling  could you please rebase this PR against the last development branch, this will fix the failing tests.

jkelling committed:
QueueGenericThreadsNonBlocking.hpp: Add include "~Unused.hpp"

jkelling committed:
EventGenericThreads: Add detail for CurrentThreadWaitFor

Rm form dev/wait/cpu.hpp, streamline trait impl

jkelling commented:
@psychocoderHPC  or @BenjaminW3  can you pleas review the latest commit I added?
https://github.com/ComputationalRadiationPhysics/alpaka/pull/959/commits/449269e2ba95928bf61a1cef954053483524156c

I found this change when rebasing my omp4 branch onto this one. Thematically belongs to this PR.
This is a change which might need to be discussed:
The template function`wait::traits::detail:currentThreadWaitForDevice` is made less general in this commit and moved into EventGenericThreads.hpp and is now `wait::traits::generic::currentThreadWaitForDevice`
1. The reason for making it less generic, is that this way the speicalizations of `CurrentThreadWaitFor<Dev>` are shorter (one line)
2. It was not used outside of CurrentThreadWaitFor<DevCpu>
3. Having something in a namespace detail generic enough to be used in any context is contradictory.
4. point 2 and 3 are in line the policy discussed here: https://github.com/ComputationalRadiationPhysics/alpaka/pull/959#discussion_r395971248

psychocoderHPC commented on include/alpaka/event/EventGenericThreads.hpp:
```diff
@@ -316,6 +317,33 @@ namespace alpaka
+                ALPAKA_FN_HOST auto currentThreadWaitForDevice(
+                    TDev const & dev
+                )
+                ->void
```
IMO for void we should not use trailing return values. Since we depend on C++14 we can skip the type if we use auto but in a void method/function it is a little bit risky if someone is adding a return.

```C++
                template<typename TDev>
￼                ALPAKA_FN_HOST void currentThreadWaitForDevice(
￼                    TDev const & dev
￼                )
```
@BenjaminW3 What do you think?

jkelling replied:
I agree, that the trailing void looks a bit silly, but it was like this in the original code.

Would be happy to change it, but this is not the only place with a trialing void, e.g. https://github.com/jkelling/alpaka/blob/queuegeneric/include/alpaka/event/EventGenericThreads.hpp#L72 , https://github.com/jkelling/alpaka/blob/queuegeneric/include/alpaka/event/EventGenericThreads.hpp#L267 , https://github.com/jkelling/alpaka/blob/queuegeneric/include/alpaka/event/EventGenericThreads.hpp#L307 , https://github.com/jkelling/alpaka/blob/queuegeneric/include/alpaka/event/EventGenericThreads.hpp#L326 ...

Should all be changed like this?

jkelling replied:
There are also trailing -> bool . What about those?

BenjaminW3 replied:
I would keep it. We use trailing return types for everything. I prefer consistency.

jkelling replied:
... code cleanup might be better as a separate PR, IMHO.

jkelling replied:
My page was not updated to show @BenjaminW3's comment when I wrote mine. Marking this as resolved for this PR.

psychocoderHPC commented:
> @psychocoderHPC or @BenjaminW3 can you pleas review the latest commit I added?
> [449269e](https://github.com/ComputationalRadiationPhysics/alpaka/commit/449269e2ba95928bf61a1cef954053483524156c)
> 
> I found this change when rebasing my omp4 branch onto this one. Thematically belongs to this PR.
> This is a change which might need to be discussed:
> The template function`wait::traits::detail:currentThreadWaitForDevice` is made less general in this commit and moved into EventGenericThreads.hpp and is now `wait::traits::generic::currentThreadWaitForDevice`
> 
> 1. The reason for making it less generic, is that this way the speicalizations of `CurrentThreadWaitFor<Dev>` are shorter (one line)
> 2. It was not used outside of CurrentThreadWaitFor
> 3. Having something in a namespace detail generic enough to be used in any context is contradictory.
> 4. point 2 and 3 are in line the policy discussed here: [#959 (comment)](https://github.com/ComputationalRadiationPhysics/alpaka/pull/959#discussion_r395971248)

IMO this change is ok, the method is not exposed to the user.
